File: algorithms/java/46_permutations/Permutation.java
class Permutation {
	public List<List<Integer>> permute(int[] nums) {
		List<List<Integer>> res = new ArrayList<>();
		if (nums == null || nums.length == 0) {
			return res;
		}
		dfs(nums, res, 0);
		return res;
	}

	private void dfs(int[] nums, List<List<Integer>> res, int index) {
		if (index == nums.length) {
			List<Integer> tmp = new ArrayList<>();
			for (int n : nums) {
				tmp.add(n);
			}
			res.add(tmp);
			return;
		}
		for (int i = index; i < nums.length; i++) {
			swap(nums, index, i);
			dfs(nums, res, index + 1);
			swap(nums, index, i);
		}
	}

	private void swap(int[] nums, int i, int j) {
		int tmp = nums[i];
		nums[i] = nums[j];
		nums[j] = tmp;
	}
}

File: algorithms/cpp/47_permutation2/Permutation.cpp
class Permutation {
public:
	vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        permute(res, nums, 0);
        return res;
    }
private:
	void permute(vector<vector<int>>& res, vector<int>& nums, int index) {
        if (index == nums.size()) {
            res.push_back(nums);
            return;
        }
        unordered_set<int> set;
        for (int i = index; i < nums.size(); ++i) {
            if (set.find(nums[i]) == set.end()) {
                set.insert(nums[i]);
                swap(nums[index], nums[i]);
                permute(res, nums, index + 1);
                swap(nums[index], nums[i]);
            }
        }
    } 
};
